File: app/src/main/java/com/mbg/mbgsupport/fragment/UtilsDemoFragment.java
package com.mbg.mbgsupport.fragment;

import android.content.Context;
import android.graphics.Color;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

import com.blued.android.module.serviceloader.Router;
import com.mbg.mbgsupport.MainActivity;
import com.mbg.mbgsupport.R;
import com.mbg.mbgsupport.router.service.IBaseService;
import com.mbg.mbgsupport.router.service.ServiceKey;
import com.mbg.mbgsupport.viewmodel.LoadingStateViewModel;
import com.mbg.module.common.core.LifecycleHandler;
import com.mbg.module.common.core.WeakHandler;
import com.mbg.module.common.core.net.manager.HttpManager;
import com.mbg.module.common.core.net.tool.HttpUtils;
import com.mbg.module.common.core.net.wrapper.response.DefaultHttpResponse;
import com.mbg.module.common.core.sharedpreference.FastSharedPreferences;
import com.mbg.module.common.util.AppUtils;
import com.mbg.module.common.util.ClickUtils;
import com.mbg.module.common.util.FileCacheUtils;
import com.mbg.module.common.util.KeyboardUtils;
import com.mbg.module.common.util.LocaleUtils;
import com.mbg.module.common.util.LogUtils;
import com.mbg.module.common.util.PermissionHelper;
import com.mbg.module.common.util.PermissionUtils;
import com.mbg.module.common.util.SpannableUtils;
import com.mbg.module.common.util.StringUtils;
import com.mbg.module.common.util.ThreadUtils;
import com.mbg.module.common.util.ToastUtils;
import com.mbg.module.common.util.UiUtils;
import com.mbg.module.common.util.UriUtils;
import com.mbg.module.common.util.consts.PermissionConsts;
import com.mbg.module.ui.activity.TerminalActivity;
import com.mbg.module.ui.fragment.BaseFragment;
import com.mbg.module.ui.view.inflate.AsyncLayoutInflatePlus;
import com.mbg.module.ui.view.inflate.OnInflateFinishedListener;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TreeSet;
import java.util.concurrent.CountDownLatch;

public class UtilsDemoFragment extends BaseFragment implements View.OnClickListener{
    private ImageView mShowImageView;
    private LoadingStateViewModel mLoadingStateViewModel;
    private WeakHandler weakHandler=new WeakHandler(this);
    private LifecycleHandler lifecycleHandler=new LifecycleHandler(this);
    private TextView weakTextView;
    private TextView lifeTextView;
    private View mGlobalView;
    private ViewGroup mRootView;


    public static void show(Context context){
        TerminalActivity.show(context, UtilsDemoFragment.class,null);
    }
    @Override
    protected int onRequestLayout() {
        return R.layout.fragment_utils_demo;
    }

    @Override
    protected void initView() {
        mRootView=findViewById(R.id.root_view);
        PermissionUtils.checkStoragePermissson(getActivity(), new PermissionUtils.PermissionCallbacks() {
            @Override
            public void onPermissionsGranted(int requestCodes, List<String> perms) {

            }

            @Override
            public void onPermissionsDenied(int requestCodes, List<String> perms) {

            }
        });

        if(!PermissionHelper.isPermissionGranted(PermissionConsts.CAMERA)){
            PermissionHelper
                    .permission(PermissionConsts.CAMERA)
                    .callback(new PermissionHelper.SimpleCallback() {
                        @Override
                        public void onGranted() {
                            LogUtils.d("onGranted");
                        }

                        @Override
                        public void onDenied() {
                            LogUtils.d("onGranted");
                        }
                    }).request();
        }

        if(!PermissionHelper.isPermissionGranted(PermissionConsts.LOCATION)){
            PermissionHelper
                    .permission(PermissionConsts.LOCATION)
                    .callback(new PermissionHelper.SimpleCallback() {
                        @Override
                        public void onGranted() {
                            LogUtils.d("onGranted");
                        }

                        @Override
                        public void onDenied() {
                            LogUtils.d("onGranted");
                        }
                    }).request();
        }

        findViewById(R.id.btn_test1).setOnClickListener(this);
        findViewById(R.id.btn_test2).setOnClickListener(this);
        findViewById(R.id.btn_test3).setOnClickListener(this);
        findViewById(R.id.btn_test4).setOnClickListener(this);
        findViewById(R.id.btn_test5).setOnClickListener(this);
        findViewById(R.id.btn_test6).setOnClickListener(this);
        findViewById(R.id.btn_test7).setOnClickListener(this);
        findViewById(R.id.btn_test8).setOnClickListener(this);
        findViewById(R.id.btn_test9).setOnClickListener(this);
        findViewById(R.id.btn_show_global).setOnClickListener(this);
        findViewById(R.id.btn_hide_global).setOnClickListener(this);
        mGlobalView=View.inflate(getActivity(),R.layout.view_global_demo,null);

        weakTextView=findViewById(R.id.btn_weak_handle);
        weakTextView.setOnClickListener(this);
        lifeTextView=findViewById(R.id.btn_life_handle);
        lifeTextView.setOnClickListener(this);

        final TextView inputRoot=findViewById(R.id.tv_test_input);

        SpannableUtils.SpannableFilter spannableFilter=new SpannableUtils.SpannableFilter();
        List<SpannableUtils.SpannableFilter> spannableFilters=new ArrayList<>();
        spannableFilter.keywords="测试";
        spannableFilter.keywordsColor= Color.RED;
        spannableFilter.isUnderLine=false;
        spannableFilter.onClickListener=new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ToastUtils.debugShow("测试");
            }
        };
        spannableFilters.add(spannableFilter);

        spannableFilter=new SpannableUtils.SpannableFilter();
        spannableFilter.keywords="键盘";
        spannableFilter.keywordsColor= Color.GRAY;
        spannableFilter.isUnderLine=true;
        spannableFilter.onClickListener=new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ToastUtils.debugShow("键盘");
            }
        };
        spannableFilters.add(spannableFilter);

        SpannableUtils.setViewSpannable(getActivity(),inputRoot,spannableFilters);

        mShowImageView=findViewById(R.id.img_show);


        KeyboardUtils.registerSoftInputChangedListener(getActivity(), new KeyboardUtils.OnSoftInputChangedListener() {
            @Override
            public void onSoftInputChanged(int height) {
                ToastUtils.debugShow("onSoftInputChanged:height="+height);
                if (height>0){
                    inputRoot.setTranslationY(-height);
                }
            }
        });

        mLoadingStateViewModel=ViewModelProviders.of(getActivity()).get("LoadingState", LoadingStateViewModel.class);
        mLoadingStateViewModel.getLoadingState().observe(getActivity(), new Observer<LoadingStateViewModel.LoadingState>() {
            @Override
            public void onChanged(LoadingStateViewModel.LoadingState loadingState) {
                if (loadingState!=null){
                    switch (loadingState){
                        case START:
                            //onDataLoadingStart();
                            break;
                        case FINISH:
                            onDataLoadingFinish();
                            break;
                    }
                }
            }
        });

        ThreadUtils.postInUIThreadDelayed(new Runnable() {
            @Override
            public void run() {
                mLoadingStateViewModel.setLoadingSate(LoadingStateViewModel.LoadingState.FINISH);
            }
        },8000);


        new AsyncLayoutInflatePlus(getActivity()).inflate(R.layout.view_async_load, null, new OnInflateFinishedListener() {
            @Override
            public void onInflateFinished(@NonNull View view, int resId, @Nullable ViewGroup parent) {
                mRootView.addView(view);
            }
        });

    }

    protected void  onDataLoadingStart(){

        HashSet hs = new HashSet();
        long ks = System.currentTimeMillis();
        for (int i=1;i<99999;i++)
            hs.add(i);
        Iterator it = hs.iterator();
        while (it.hasNext())
            it.next();
        boolean add=hs.add(1);
        hs.remove(5555);
        hs.clear();
        long js = System.currentTimeMillis();


        //LinkedHashSet效率
        LinkedHashSet lh = new LinkedHashSet();
        long ks1 = System.currentTimeMillis();
        for (int i=1;i<99999;i++)
            lh.add(i);


        Iterator it1 = lh.iterator();
        while (it1.hasNext())
            it1.next();
        add=lh.add(1);
        lh.remove(5555);
        lh.clear();
        long js1 = System.currentTimeMillis();


        //TreeSet效率
        TreeSet ts = new TreeSet();
        long ks2 = System.currentTimeMillis();
        for (int i=1;i<99999;i++)
            ts.add(i);


        Iterator it2 = ts.iterator();
        while (it2.hasNext())
            it2.next();
        add=ts.add(1);
        add=ts.add(0);
        ts.remove(5555);
        ts.clear();
        long js2 = System.currentTimeMillis();

        LogUtils.d("HashSet共花费时间："+(js-ks)+"ms");
        LogUtils.d("LinkedHashSet共花费时间："+(js1-ks1)+"ms");
        LogUtils.d("TreeSet共花费时间："+(js2-ks2)+"ms");

    }

    protected void  onDataLoadingFinish(){
        LogUtils.d("LoadingStateViewModel:onDataLoadingFinish");
    }

    @Override
    public void onClick(View view) {
        int viewId=view.getId();
        if(ClickUtils.isFastDoubleClick(viewId)){
            return;
        }
        switch (viewId){
            case R.id.btn_test1:
                onTest1();
                FastSharedPreferences.get("FSP_DATA_USER").edit().putBoolean("keyBool",true);
                FastSharedPreferences.get("FSP_DATA_USER").edit().putString("keyString","btn_test1").apply();
                break;
            case R.id.btn_test2:
                FastSharedPreferences.get("FSP_DATA_USER").edit().putBoolean("keyBool",false);
                FastSharedPreferences.get("FSP_DATA_USER").edit().putString("keyString","btn_test2").apply();
                onTest2();
                break;
            case R.id.btn_test3:
                ToastUtils.show("keyBool="+ FastSharedPreferences.get("FSP_DATA_USER").getBoolean("keyBool",false));
               // onTest3();
                break;
            case R.id.btn_test4:
                ToastUtils.show("keyString="+ FastSharedPreferences.get("FSP_DATA_USER").getString("keyString",""));
                //onTest4();
                break;
            case R.id.btn_test5:
                onTest5();
                break;
            case R.id.btn_test6:
                onTest6();
            case R.id.btn_test7:
                onTest7();
                break;
            case R.id.btn_test8:
                onTest8();
                break;
            case R.id.btn_test9:
                onTest9();
                break;
            case R.id.btn_weak_handle:
                for (int i=0;i<1000;i++) {
                    weakHandler.postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            weakTextView.setText("我3秒前被点击了");
                            weakTextView.setTextColor(Color.GREEN);
                        }
                    }, 10000);
                }
                break;
            case R.id.btn_life_handle:
                lifecycleHandler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        lifeTextView.setText("我3秒前被点击了");
                        lifeTextView.setTextColor(Color.RED);
                    }
                },3000);
                break;
            case R.id.btn_show_global:
                UiUtils.showGlobal(getActivity(), mGlobalView, R.id.view_global, 500, new UiUtils.ViewCallback() {
                    @Override
                    public boolean onView(View view) {
                        return true;//返回true,表示不重新加入
                    }
                });
                break;
            case R.id.btn_hide_global:
                UiUtils.hideGlobal(getActivity(),R.id.view_global,500);
                break;
        }

    }

     private void onTest1(){

       /* String path= Environment.getExternalStorageDirectory().getPath()+"/DCIM/Image/a001.png";
        Uri uri= UriUtils.getUriFromFile(path);
        LogUtils.v(uri.toString());
        mShowImageView.setImageURI(uri);*/
         mLoadingStateViewModel.setLoadingSate(LoadingStateViewModel.LoadingState.START);
     }
    private void onTest2(){
        new Thread(new Runnable() {
            @Override
            public void run() {
                IBaseService service= Router.getService(IBaseService.class, ServiceKey.KEY_SERVICE);
                service.doInBackground();
            }
        }).start();
        mLoadingStateViewModel.setLoadingSate(LoadingStateViewModel.LoadingState.FINISH);

    }
    private void onTest3() {
        String key="Test";
        String str = FileCacheUtils.getContent(key);
        if (StringUtils.isEmpty(str)) {
            FileCacheUtils.saveContent(key,"Fisrt save,大中国。。");
            ToastUtils.show("Cache is empty!");
        } else {
            ToastUtils.show("Cache:" + str);
        }
    }
    private void onTest4(){

        DefaultHttpResponse response=new DefaultHttpResponse() {
            @Override
            protected void onUpdate(String data) {
                super.onUpdate(data);
                LogUtils.e("onUpdate:"+data);
            }

            @Override
            public void onUIStart() {
                super.onUIStart();
                LogUtils.e("onUIStart");
            }

            @Override
            public void onUICache(String data) {
                super.onUICache(data);
                LogUtils.e("onUICache:"+data);
            }

            @Override
            public void onUIError(Exception error) {
                super.onUIError(error);
                LogUtils.e("onUIError:"+error.getMessage());
            }

            @Override
            public void onUIFinish() {
                super.onUIFinish();
                LogUtils.e("onUIFinish");
            }
        };

        String url="http://tcc.taobao.com/cc/json/mobile_tel_segment.htm?tel=18701539645";
        HttpManager.get(url,response).addHeader(HttpUtils.buildBaseHeader(false)).execute();
    }


    private void onTest5(){
        LocaleUtils.setLocale(getActivity(),new Locale(LocaleUtils.Language.ZH.getCode(),LocaleUtils.CountryArea.China.getCode()));
        AppUtils.rebootApplication(MainActivity.class);

    }
    private void onTest6(){
        LocaleUtils.setLocale(getActivity(),new Locale(LocaleUtils.Language.EN.getCode(),LocaleUtils.CountryArea.America.getCode()));
        AppUtils.rebootApplication(MainActivity.class);
    }

    private void onTest7(){
        KeyboardUtils.showSoftInput(getActivity());
    }
    private void onTest8(){
       KeyboardUtils.hideSoftInput(getActivity());
    }
    private void onTest9(){
        //showContent(ImageLoaderFragment.class);
      Map<Thread,StackTraceElement[]> map= Thread.getAllStackTraces();
        for(Map.Entry<Thread,StackTraceElement[]> entry : map.entrySet()){
            Thread mapKey = entry.getKey();
            System.out.println("Thread:"+mapKey.getName()+" ,isAlive:"+mapKey.isAlive()+" ,ThreadID:"+mapKey.getId());
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        weakTextView=null;
        lifeTextView=null;
    }
}

File: app/src/main/java/com/mbg/mbgsupport/fragment/UtilsDemoFragment.kt
package com.mbg.mbgsupport.fragment

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.view.animation.OvershootInterpolator
import android.widget.LinearLayout
import androidx.lifecycle.ViewModelProviders
import com.blued.android.module.serviceloader.Router
import com.mbg.mbgsupport.MainActivity
import com.mbg.mbgsupport.R
import com.mbg.mbgsupport.databinding.FragmentUtilsDemoBinding
import com.mbg.mbgsupport.demo.kotlin.mvp.DemoPresenter
import com.mbg.mbgsupport.router.service.IBaseService
import com.mbg.mbgsupport.router.service.ServiceKey
import com.mbg.mbgsupport.viewmodel.LoadingStateViewModel
import com.mbg.mbgsupport.viewmodel.LoadingStateViewModel.LoadingState
import com.mbg.module.common.core.LifecycleHandler
import com.mbg.module.common.core.WeakHandler
import com.mbg.module.common.core.manager.CoroutineManager.Companion.get
import com.mbg.module.common.core.net.manager.HttpManager
import com.mbg.module.common.core.net.tool.HttpUtils
import com.mbg.module.common.core.net.wrapper.response.DefaultHttpResponse
import com.mbg.module.common.core.sharedpreference.FastSharedPreferences
import com.mbg.module.common.util.*
import com.mbg.module.common.util.SpannableUtils.SpannableFilter
import com.mbg.module.common.util.consts.PermissionConsts
import com.mbg.module.ui.kotlin.activity.PhoneActivity
import com.mbg.module.ui.mvp.kotlin.MvpFragment
import com.mbg.module.ui.view.inflate.AsyncLayoutInflatePlus
import java.util.*

class UtilsDemoFragment : MvpFragment<DemoPresenter,FragmentUtilsDemoBinding>() {
    private val weakHandler = WeakHandler(this)
    private val lifecycleHandler = LifecycleHandler(this)
    private var mLoadingStateViewModel:LoadingStateViewModel?=null

    override fun onPresenterResult(resultCode: Int, vararg params: Any?) {

    }

    override fun onLoadMoreView(isHasMore: Boolean) {

    }

    override fun onLoadFinish() {

    }

    override fun onInitView(savedInstanceState: Bundle?) {
        super.onInitView(savedInstanceState)
        initView()
    }

    override fun initView() {
        PermissionUtils.checkStoragePermissson(activity,
            object : PermissionUtils.PermissionCallbacks {
                override fun onPermissionsGranted(requestCodes: Int, perms: List<String>) {}
                override fun onPermissionsDenied(requestCodes: Int, perms: List<String>) {}
            })
        if (!PermissionHelper.isPermissionGranted(PermissionConsts.CAMERA)) {
            PermissionHelper.permission(PermissionConsts.CAMERA)
                .callback(object : PermissionHelper.SimpleCallback {
                    override fun onGranted() {
                        LogUtils.d("onGranted")
                    }

                    override fun onDenied() {
                        LogUtils.d("onGranted")
                    }
                }).request()
        }
        if (!PermissionHelper.isPermissionGranted(PermissionConsts.LOCATION)) {
            PermissionHelper.permission(PermissionConsts.LOCATION)
                .callback(object : PermissionHelper.SimpleCallback {
                    override fun onGranted() {
                        LogUtils.d("onGranted")
                    }

                    override fun onDenied() {
                        LogUtils.d("onGranted")
                    }
                }).request()
        }
        mViewBinding?.run {

            val inputRoot = tvTestInput
            var spannableFilter = SpannableFilter()
            val spannableFilters: MutableList<SpannableFilter> = ArrayList()
            spannableFilter.keywords = "测试"
            spannableFilter.keywordsColor = Color.RED
            spannableFilter.isUnderLine = false
            spannableFilter.onClickListener = View.OnClickListener { ToastUtils.debugShow("测试") }
            spannableFilters.add(spannableFilter)
            spannableFilter = SpannableFilter()
            spannableFilter.keywords = "键盘"
            spannableFilter.keywordsColor = Color.GRAY
            spannableFilter.isUnderLine = true
            spannableFilter.onClickListener = View.OnClickListener { ToastUtils.debugShow("键盘") }
            spannableFilters.add(spannableFilter)
            SpannableUtils.setViewSpannable(activity, inputRoot, spannableFilters)

            KeyboardUtils.registerSoftInputChangedListener(requireActivity()) { height ->
                ToastUtils.debugShow("onSoftInputChanged:height=$height")
                if (height > 0) {
                    inputRoot.translationY = -height.toFloat()
                }
            }

            mLoadingStateViewModel = ViewModelProviders.of(requireActivity()).get("LoadingState", LoadingStateViewModel::class.java)
            mLoadingStateViewModel?.loadingState?.observe(requireActivity(), { loadingState ->
                if (loadingState != null) {
                    when (loadingState) {
                        LoadingState.START -> {
                        }
                        LoadingState.FINISH -> onDataLoadingFinish()
                    }
                }
            })
            ThreadUtils.postInUIThreadDelayed({ mLoadingStateViewModel!!.setLoadingSate(LoadingState.FINISH) }, 8000)
            AsyncLayoutInflatePlus(requireActivity()).inflate(R.layout.view_async_load, null) { view, resId, parent -> root.addView(view) }

            val mGlobalView = View.inflate(activity, R.layout.view_global_demo, null)

            btnTest1.setOnClickListener {
                onTest1()
                FastSharedPreferences.get("FSP_DATA_USER").edit().putBoolean("keyBool", true)
                FastSharedPreferences.get("FSP_DATA_USER").edit()
                    .putString("keyString", "btn_test1").apply()
            }
            btnTest2.setOnClickListener {
                FastSharedPreferences.get("FSP_DATA_USER").edit().putBoolean("keyBool", false)
                FastSharedPreferences.get("FSP_DATA_USER").edit()
                    .putString("keyString", "btn_test2").apply()
                onTest2()
            }

            btnTest3.setOnClickListener {
                ToastUtils.show("keyBool=" + FastSharedPreferences.get("FSP_DATA_USER")
                    .getBoolean("keyBool", false))
            }

            btnTest4.setOnClickListener {
                ToastUtils.show("keyString=" + FastSharedPreferences.get("FSP_DATA_USER")
                    .getString("keyString", ""))
            }

            btnTest5.setOnClickListener {

                onTest5()
            }

            btnTest6.setOnClickListener {

            }

            btnTest7.setOnClickListener {
                onTest7()
            }

            btnTest8.setOnClickListener {
                onTest8()
            }

            btnTest9.setOnClickListener {
                onTest9()
            }

            btnShowGlobal.setOnClickListener {
                UiUtils.showGlobal(requireActivity(), mGlobalView, R.id.view_global, 500) {
                    true //返回true,表示不重新加入
                }
                ThreadUtils.postInUIThreadDelayed({
                    UiUtils.hideGlobal(requireActivity(),
                        R.id.view_global,
                        500)
                }, 2000)
            }

            btnHideGlobal.setOnClickListener {
                UiUtils.hideGlobal(requireActivity(), R.id.view_global, 500)
            }

            btnAr.setOnClickListener {
                onAR()
            }

            btnCoroutine.setOnClickListener {
                var i = 0
                while (i < 1000) {
                    get().start(Runnable {
                        try {
                            Thread.sleep(400)
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    })
                    i++
                }
            }

            btnShowAnim.setOnClickListener {
                showAnimal()
            }

            btnHideAnim.setOnClickListener {
                hideAnimal()
            }

            btnWeakHandle.setOnClickListener {
                var i = 0
                while (i < 1000) {
                    weakHandler.postDelayed({
                        btnWeakHandle.text = "我3秒前被点击了"
                        btnWeakHandle.setTextColor(Color.GREEN)
                    }, 10000)
                    i++
                }
            }

            btnLifeHandle.setOnClickListener {
                lifecycleHandler.postDelayed({
                    btnLifeHandle.text = "我3秒前被点击了"
                    btnLifeHandle.setTextColor(Color.RED)
                }, 3000)
            }

            btnAnimal.setOnClickListener {

            }


        }






    }

    private fun onDataLoadingFinish() {
        LogUtils.d("LoadingStateViewModel:onDataLoadingFinish")
    }


    private fun showAnimal() {
        mViewBinding?.run {
            val animator = ValueAnimator.ofInt(-btnAnimal.height, 0)
            animator.duration = 300
            animator.interpolator = OvershootInterpolator()
            val layoutParams = btnAnimal.layoutParams as LinearLayout.LayoutParams
            animator.addUpdateListener { valueAnimator ->
                layoutParams.topMargin = valueAnimator.animatedValue as Int
                btnAnimal.layoutParams = layoutParams
            }
            animator.start()
        }
    }

    private fun hideAnimal() {
        mViewBinding?.run {
            val animator = ValueAnimator.ofInt(0, -btnAnimal.height)
            animator.duration = 300
            animator.interpolator = OvershootInterpolator()
            val layoutParams = btnAnimal.layoutParams as LinearLayout.LayoutParams
            animator.addUpdateListener { valueAnimator ->
                layoutParams.topMargin = valueAnimator.animatedValue as Int
                btnAnimal.layoutParams = layoutParams
            }
            animator.start()
        }
    }

    private fun onTest1() {

        mLoadingStateViewModel!!.setLoadingSate(LoadingState.START)
    }

    private fun onTest2() {
        Thread {
            val service = Router.getService(IBaseService::class.java, ServiceKey.KEY_SERVICE)
            service.doInBackground()
        }.start()
        mLoadingStateViewModel!!.setLoadingSate(LoadingState.FINISH)
    }

    private fun onTest3() {
        val key = "Test"
        val str = FileCacheUtils.getContent(key)
        if (StringUtils.isEmpty(str)) {
            FileCacheUtils.saveContent(key, "Fisrt save,大中国。。")
            ToastUtils.show("Cache is empty!")
        } else {
            ToastUtils.show("Cache:$str")
        }
    }

    private fun onTest4() {
        val response: DefaultHttpResponse = object : DefaultHttpResponse() {
            override fun onUpdate(data: String) {
                super.onUpdate(data)
                LogUtils.e("onUpdate:$data")
            }

            override fun onUIStart() {
                super.onUIStart()
                LogUtils.e("onUIStart")
            }

            override fun onUICache(data: String) {
                super.onUICache(data)
                LogUtils.e("onUICache:$data")
            }

            override fun onUIError(error: Exception) {
                super.onUIError(error)
                LogUtils.e("onUIError:" + error.message)
            }

            override fun onUIFinish() {
                super.onUIFinish()
                LogUtils.e("onUIFinish")
            }
        }
        val url = "http://tcc.taobao.com/cc/json/mobile_tel_segment.htm?tel=18701539645"
        HttpManager.get(url, response).addHeader(HttpUtils.buildBaseHeader(false)).execute()
    }

    private fun onTest5() {
        LocaleUtils.setLocale(activity,
            Locale(LocaleUtils.Language.ZH.code, LocaleUtils.CountryArea.China.code))
        AppUtils.rebootApplication(MainActivity::class.java)
    }

    private fun onTest6() {
        LocaleUtils.setLocale(activity,
            Locale(LocaleUtils.Language.EN.code, LocaleUtils.CountryArea.America.code))
        AppUtils.rebootApplication(MainActivity::class.java)
    }

    private fun onAR() {
        LocaleUtils.setLocale(activity,
            Locale(LocaleUtils.Language.AR.code, LocaleUtils.CountryArea.Arab.code))
        AppUtils.rebootApplication(MainActivity::class.java)
    }

    private fun onTest7() {
        KeyboardUtils.showSoftInput(requireActivity())
    }

    private fun onTest8() {
        KeyboardUtils.hideSoftInput(requireActivity())
    }

    private fun onTest9() { //showContent(ImageLoaderFragment.class);
        val map = Thread.getAllStackTraces()
        for ((mapKey) in map) {
            println("Thread:" + mapKey.name + " ,isAlive:" + mapKey.isAlive + " ,ThreadID:" + mapKey.id)
        }
    }



    override fun onDestroyView() {
        super.onDestroyView()

        get().clear()
    }

    companion object {
        @JvmStatic
        fun show(context: Context) {
            PhoneActivity.show(context, UtilsDemoFragment::class.java, null)
        }
    }
}